DataDog support added

Ramizjon opened the pull request:
Added reporting task, that reports NiFi flow and JVM metrics to DataDog service. It also reports data to DataDog from every processor. All metrics are represented as Gauges. The name of each metric for processors has following format: “nifi.<processor_name>.<metric_name>”. In order to use this module, you need to have DataDog agent installed on your machine.

joewitt commented:
@Ramizjon this is really cool.  have you already opened a JIRA for it?  Please do so and/or link the commit message to it.  More information on contribution process please see here https://cwiki.apache.org/confluence/display/NIFI/Contributor+Guide

JPercivall commented:
Hello @Ramizjon, I'm gonna review this for you. Couple admin things to start.

First, the versions of the datadog poms are "0.6.1". These should be changed to "1.0.0-SNAPSHOT". Second, the datadog bundle should be listed in the nifi-nar-bundle pom as a module. Also the classes you create should have the ASF license at the top. [Here](https://github.com/apache/nifi/blob/c3d54ab7246fba5ea1432d949c79f97f0e97f25c/nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/GetKafka.java#L1-L1) is an example. The offending files are listed below. Lastly the nar should include a License in addition to the notice.


Unapproved licenses:

  /Users/jpercivall/projects/nifi/nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java
  /Users/jpercivall/projects/nifi/nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DDMetricRegistryBuilder.java
  /Users/jpercivall/projects/nifi/nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/test/java/org/apache/nifi/reporting/datadog/TestDataDogReportingTask.java
  /Users/jpercivall/projects/nifi/nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/test/java/org/apache/nifi/reporting/datadog/TestMetricsService.java

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/pom.xml:
```diff
@@ -0,0 +1,86 @@
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>19.0</version>
+        </dependency>
```
These three deps are listed under the "test dependencies" section but are not scoped as such. Can they be made scope "test"?

Ramizjon replied:
Not all of them are "test". Will fix it.

JPercivall commented:
Also the nar needs to be added as a dependency of the nifi-assembly nar and the license/notice information added to the nifi-assembly license/notice if there is any new deps that need to be added.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/resources/docs/org.apache.nifi.reporting.ambari.AmbariReportingTask/additionalDetails.html:
```diff
@@ -0,0 +1,56 @@
+    <body>
+        <h2>DataDogReportingTask</h2>
+
+        <p>This ReportingTask sends the following metrics to DataDog:</p>
```
I don't know much about DataDog, but will a user always care about all of these stats? Wondering if it should be configurable.

Ramizjon replied:
User will add to his DataDog dashboard only that metrics that he is interested in. Other will not be shown, but they will exist, and he'll be able to include every of them at any moment.

irabinovitch replied:
If we can determine some base metrics that apply to everyone, I'd be happy to help you create a default dashboard in the Datadog service.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,185 @@
+    public void onTrigger(ReportingContext context) {
+        final ProcessGroupStatus status = context.getEventAccess().getControllerStatus();
+        final String reportingPeriod = context.getProperty(REPORTING_PERIOD)
+                .evaluateAttributeExpressions().getValue();
```
You are running "evaluateAttributeExpressions" on a property that doesn't support expression language.

JPercivall replied:
Same with the other properties too.

Ramizjon replied:
Will fix that.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,185 @@
+    private String statusId;
+    private ConcurrentHashMap<String, AtomicDouble> metricsMap;
+    private volatile VirtualMachineMetrics virtualMachineMetrics;
+    private Logger logger = LoggerFactory.getLogger(getClass().getName());
```
The framework provides a logger to use with the "getLogger()" method.

Ramizjon replied:
Ok, will fix that.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DDMetricRegistryBuilder.java:
```diff
@@ -0,0 +1,71 @@
+        UdpTransport udpTransport = new UdpTransport.Builder().build();
+        DatadogReporter reporter =
+                DatadogReporter.forRegistry(metricRegistry)
+                        .withHost(InetAddress.getLocalHost().getHostName())
```
Why can't this be made report to another host?

Ramizjon replied:
It's a predefined tag that is called "host". Again, we report metrics to DataDog agent, that is installed locally, and it reports these metrics to DataDog

JPercivall commented:
@Ramizjon could you explain the flow of logic that is going on? I think you're mixing up what each of the ReportingTask methods do and not letting the framework help.

The flow of logic from the NiFi side is, the user will configure a period for the Reporting task to run. When they "start" the task the "@OnScheduled" method will be run. After that the framework will keep calling the "ontrigger" method, with the period the user set, in order to report. 

I believe what you want to do is to initialize the "datadogReporter" in the OnScheduled using the user defined properties and then in the onTrigger call the "report" method of "datadogReporter".

JPercivall commented:
Also doesn't DataDog use API keys in order to communicate? Shouldn't that be configured in the ReportingTask?

Ramizjon commented:
@JPercivall Thank you for your comments. 
About two last notes: in order to report to DataDog the DropWizard metrics "Gauges" are used (A gauge is an instantaneous measurement of a value). They work in such way, that you register some metric and specify where to take the value of every metric from (line with `ConcurrentHashMap<String, AtomicDouble> metricsMap`). After that, DropWizard framework gets metric values from that structure with specified period of time.

> Also doesn't DataDog use API keys in order to communicate? Shouldn't that be configured in the ReportingTask?
In order do deliver metrics to DataDog, the DataDog agent is used. It is the best way do to so. It collects metrics and events and reports them to DataDog. This approach doesn't require any of  API keys, but only agent installed.

JPercivall commented:
@Ramizjon, the gauges make sense but the last part, "DropWizard framework gets metric values from that structure with specified period of time", doesn't. It seems (looking at this line of code[1]) that the "start" method of the DatadogReporter is causing the reporting with a specified period of time. If you just used one of the "report"[2] methods every onTrigger call you would achieve what I am suggesting.

Installing an agent may be very simple, I don't like the idea of requiring a user to install something on the box in order to work (the person configuring the flow may not have access to install locally). I would very much prefer go the extra mile to make this work without an agent locally. You mention that it's the "best way", this means you're able to report without having an agent locally? 

[1] https://github.com/apache/nifi/pull/655/files#diff-3f6f231dd20c75c7f8830cc0fd156770R55
[2] https://github.com/coursera/metrics-datadog/blob/master/metrics-datadog/src/main/java/org/coursera/metrics/datadog/DatadogReporter.java#L71

Ramizjon commented:
@JPercivall Yes, i've already did it in the way you recommend, by calling report() method every onTrigger() call.

DataDog agent is very commonly used approach of sending metrics to DataDog. Firstly, it goes about aggregation. It is used to aggregate many data points from different sources (not only from NiFi) and send them. So if user is using DataDog for a lot of purposes it's obvious that it is more efficient to use such unified module as agent. Secondly, it works much faster than direct approach, because of accepting custom application metrics points over UDP.  The application won’t stop its actual work to wait for a response from the metrics server, which is very important if the metrics server is down or inaccessible. Thirdly, DataDog provides single-line install commands for every OS, so it is absolutely painless. However i suggest to add direct approach as additional feature, for users that don't want to install an Agent.

Ramizjon commented:
@JPercivall However, if you strongly recommend using direct approach, i can implement it.

JPercivall commented:
@Ramizjon if there are signifiant improvements to do the local agent vs direct, I would suggest making it configurable. That way the user can use the Reporting task without needs access to the box but also make the option available for the users that need the extra performance.

Ramizjon commented:
@JPercivall totally agree with you.

irabinovitch commented:
Making it configurable makes sense. That being said the recommended method would be to submit metrics via metrics-datadog's statsd implementation.  This ensures the metrics are emitted asynchronously, and are buffered should connectivity between NiFi and Datadog be interrupted any reason.

Ramizjon commented:
@JPercivall PR updated

Ramizjon commented:
@JPercivall , the type of DataDog transport is now configurable, as we've discussed, with way to specify API key. Changed logic of reporting, now it's reporting metrics to DataDog every onTrigger call.  Added licenses and fixed dependencies. Some more refactoring is done.

JPercivall commented:
@Ramizjon, thank you for updating the PR. I will probably be able to review it this weekend. Could you rebase against the latest master?

Ramizjon commented:
@JPercivall Thank you for your attention. I've synced with latest master.

JPercivall commented on nifi-assembly/NOTICE:
```diff
@@ -1,4 +1,4 @@
-Apache NiFi
+wApache NiFi
```
typo

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-nar/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,585 @@
+    THE SOFTWARE.
+
+
+This product bundles 'jsonlint' which is available under an MIT license.
```
There appears to be a lot of UI related items in this License, is that correct?

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DDMetricRegistryBuilder.java:
```diff
@@ -0,0 +1,103 @@
+            metricRegistry = new MetricRegistry();
+
+        if (name == null) {
+            name = RandomStringUtils.randomAlphanumeric(8);
```
I'm confused where the "name" field comes into play. I don't see it read anywhere except to check if it's null here.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,223 @@
+public class DataDogReportingTask extends AbstractReportingTask {
+
+    static final AllowableValue DATADOG_AGENT = new AllowableValue("DataDog Agent", "DataDog UDP Agent",
+            "Metrics will be sent via locally installed DataDog agent over UDP");
```
More of a DataDog question, isn't data loss inherent with UDP? I'm surprised UDP is the preferred method to get the data to the Agent.

JPercivall replied:
Also it should be made more explicit that this option requires the user manually installs an agent and that it is not done automatically.

irabinovitch replied:
UDP only used when posting the local agent. The agent in turn forward over HTTP to the Datadog ingest APIs.

Locally you're pushing to datadog via statsd which is a fairly standard  metrics protocol over UDP.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,223 @@
+            .description("Transport through which metrics will be sent to DataDog")
+            .required(true)
+            .allowableValues(DATADOG_AGENT, DATADOG_HTTP)
+            .defaultValue(DATADOG_AGENT.getValue())
```
I'd prefer DATADOG_HTTP be made the default value since it does not require an external thing to be installed on the local system.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,223 @@
+            .description("Environment, dataflow is running in. " +
+                    "This property will be included as metrics tag.")
+            .required(true)
+            .expressionLanguageSupported(false)
```
Kinda contrary to a previous comment, I'd suggest making this support EL (and then adding back in the evaluation). This is because with EL it allows the user to access environment variables which may be of use. Sorry about that.

This should probably also be done to METRICS_PREFIX.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,223 @@
+    private String metricsPrefix;
+    private String environment;
+    private String statusId;
+    private String apiKey;
```
This variable is assigned but never accessed.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,223 @@
+
+    @Override
+    public void onTrigger(ReportingContext context) {
+        final ProcessGroupStatus status = context.getEventAccess().getControllerStatus();
```
The input/output port statuses in this group (or the child groups) are not accounted for.

JPercivall replied:
Also doesn't include connections. Maybe a new property that allows the user to specify which stats they care about (processor, ports, connections, JVM, dataflow, etc.)

Ramizjon replied:
@JPercivall well, if user cares about more than one statistic and wants to select several of them, how to implement such property most efficiently?

JPercivall replied:
Thinking about it a bit more the best way would be to create a custom UI for the processor and enable the user to pick and chose sections (those mentioned before) and subsections (FF in/out, etc.) but I think that could be left as an improvement at a later time.

For now it is probably fine to just include all the relevant sections/subsections.

Ramizjon replied:
@JPercivall port and connection statuses in this group are empty, so could you please advice how can these statuses be retrieved from this place? Thanks.

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/metrics/MetricsService.java:
```diff
@@ -0,0 +1,120 @@
+        metrics.put(MetricNames.FLOW_FILES_SENT, String.valueOf(status.getOutputCount()));
+        metrics.put(MetricNames.BYTES_READ, String.valueOf(status.getInputBytes()));
+        metrics.put(MetricNames.BYTES_WRITTEN, String.valueOf(status.getOutputBytes()));
+        metrics.put(MetricNames.ACTIVE_THREADS, String.valueOf(status.getActiveThreadCount()));
```
I'd suggest adding "status.getProcessingNanos()" as well.

irabinovitch commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java:
```diff
@@ -0,0 +1,223 @@
+@CapabilityDescription("Publishes metrics from NiFi to datadog")
+public class DataDogReportingTask extends AbstractReportingTask {
+
+    static final AllowableValue DATADOG_AGENT = new AllowableValue("DataDog Agent", "DataDog UDP Agent",
```
s/DataDog/Datadog/

JPercivall commented on nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/metrics/MetricsService.java:
```diff
@@ -0,0 +1,120 @@
+    //general metrics for whole dataflow
+    public Map<String, String> getDataFlowMetrics(ProcessGroupStatus status) {
+        final Map<String, String> metrics = new HashMap<>();
+        metrics.put(MetricNames.FLOW_FILES_RECEIVED, String.valueOf(status.getFlowFilesReceived()));
```
Why parse everything as a String here and then parse everything again in the updateMetrics method but as a double?

JPercivall commented on nifi-assembly/LICENSE:
```diff
@@ -1487,3 +1487,56 @@ This project bundles 'Jython' which is available under a Python Software Foundat
     NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
     SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+This project bundles ‘Mockito’ which is available under an MIT license
```
This should not be added to the Assembly License. Mockito is a test dep only and thus should not be included in the Licensing of the binary distribution.

Ramizjon replied:
@JPercivall okay, understood. Will fix this and other noted parts.

Ramizjon commented:
@JPercivall PR updated. I will be grateful for the further review.

Ramizjon commented:
@JPercivall could you please take a look on PR when you'll have time? Thanks

JPercivall commented:
Hey @Ramizjon, I will take a look now. Sorry I haven't gotten a chance to look at it earlier, been very busy getting out 1.0.0-Beta

JPercivall commented:
@Ramizjon, I am attempting to test but I think there is something wrong with my Datadog account itself because I tried connecting to it with both NiFi and the OSX Agent and it doesn't get past the initial "Install your first Datadog Agent". 

I reached out to @irabinovitch (works for Datadog doing community building) out of band in order to see what's wrong with my account and will comment again when I get things working.

In the mean time, could you squash your commits so that it'll be easier to merge them into master? We like to have one commit per ticket/PR in order make the history easier to understand.

Thanks again @Ramizjon for taking the time work to work on this!

JPercivall commented:
Hey @Ramizjon, it looks like your squash didn't quite go right. I think you somehow squashed all the commits ever. That commit says  it has "4,279 changed files with 19,598 additions and 0 deletions." Honestly I'm not really sure how you managed to do that. You're supposed to squash all the commits *you* did into one commit.

Functional testing the Reporting Task, I got my OSX Datadog agent to report but the Reporting Task doesn't seem to report. I tried both HTTP and Agent configurations but neither seem to work and there are no errors being reported. Also I see in the logs the output of the report but I see nothing in my Datadog dashboard. Any suggestions?

Ramizjon commented:
@JPercivall yes, i faced some difficulties while squashing all commits into one. Still trying to do that in valid way. 

So, you have added some widgets on the dashboard and specified metrics for them, but charts are still empty? I've tried to do the same right now, and it works..

![image](https://cloud.githubusercontent.com/assets/3027035/17558623/e9900106-5f23-11e6-9474-8dbce0b35237.png)

Ramizjon commented:
@JPercivall Trying to squash all commits seems to be unsuccessful, as it causes squashing upstream commits as well and generally is painful. Probably i'll create new repo with current state of the working copy and open new PR from there with only one commit?

JPercivall commented:
@Ramizjon don't need to create a whole repo. You can probably achieve it this way (first do "git fetch --all" to make sure you're up to date):

1: branch off of your master branch (the one with the Datadog changes) to a new branch
2: do a soft reset far enough back that you can later fast-forward to the latest apache master commit
3: commit just the changes you want to make (may need to do a "git reset HEAD --hard" after committing your changes to get rid of any random changes you don't want to commit)
4: do "git rebase origin/master". 

This should fast forward your Datadog commit ahead of the latest of the master branch. Also you may want to check out this documentation on how to re-write history in Git[1]. If you get too messed up you can always use "reflog" to rewind. Let me know if you have any questions.

Also I didn't realize I needed to explicitly create a dashboard, I thought it would at least show a notification that I connected (especially when using HTTP). I will re-test once we get everything squashed.

[1] https://www.atlassian.com/git/tutorials/rewriting-history

Ramizjon commented:
@JPercivall Thank you for you answer. Just a few moments to clarify:

> 2: do a soft reset far enough back that you can later fast-forward to the latest apache master commit
Should i soft reset to my very first commit, by specifying it's hash?

And after performing all the steps you've described, should i do some extra steps for squashing or that will be enough? 

Sorry, i am not very confident with extra git features, that's why i want to clarify all to do everything right.

JPercivall commented:
No problem, happy to help get it right and learn.

When you do "git status" on the master branch it will tel you "you are X number of commits ahead of origin/master and Y number of commits behind origin/master" (may be other way around, did it from memory). This just means that you start differing from what origin/master has X commits ago and there are Y more commits beyond that point in origin/master.

So you will do "git reset HEAD~X" where X is the number of commits you are ahead. This will bring you back to a common point with origin/master but leave the changes that were made as unstaged/uncommitted changes.

After performing the steps you should be at a point where it says (when running "git status") "you are 1 commit ahead of master". This 1 commit should have all the changes you made for the Datadog implementation (and only those changes).

The "reflog" is a history of all the git commands you do. So if you get too mixed up, you can use reflog to rewind back to a certain command (ie. before you started everything).

JPercivall commented:
@Ramizjon, another great resource for learning Git is this "Explain Git with D3" page[1]. It visually and interactively explains the different commands.


[1] https://onlywei.github.io/explain-git-with-d3/

Ramizjon commented:
Thank you a lot for detailed explanation. But unfortunately i still face problems while doing that, so i'll describe what i'm doing step-by-step.

1) I'm on master. Do `git fetch --all`. Then `git status` says 

> Your branch is up-to-date with 'origin/master'.
> nothing to commit, working directory clean

2) I do `git checkout -b ramizjon_datadog`. Switched to new branch.
3) When i do `git show HEAD~8` i see "added datadog bundle" commit message. Well, this is my first commit.
4) Do `git reset --soft HEAD~8`. `git status` shows me hundreds of files.
5) I do `git reset` to unstage all the files. 
6) I add all my files(that i've ever created or modified) via `git add <filename>`.
7) I do `git commit -m "squashing commits"`.
8) As you've recommended i do `git reset HEAD --hard` to get rid of changes i don't want to commit. I see 

> HEAD is now at da9ceac squashing commits

9) I do `git rebase origin/master`. And see 

> error: The following untracked working tree files would be overwritten by checkout:
>....

10) Ok, i add all files to index.
11) I do `git rebase origin/master` again. I see some messages with content like this.

> CONFLICT (rename/delete): nifi-nar-bundles/nifi-ambari-bundle/nifi-ambari-reporting-task/src/main/java/org/apache/nifi/reporting/ambari/api/MetricBuilder.java deleted in squashing commits and renamed in HEAD. Version HEAD of nifi-nar-bundles/nifi-ambari-bundle/nifi-ambari-reporting-task/src/main/java/org/apache/nifi/reporting/ambari/api/MetricBuilder.java left in tree.
> error: Failed to merge in the changes.
> Patch failed at 0001 squashing commits

At this step i realize that something went wrong. Could you please tell at what step has occured the mistake? Sorry for taking your time while resolving this issue...

JPercivall commented:
It looks like your "origin" is not the apache or github repo so your not comparing to the correct master commit. Could you run this "git remote -v"? 

That will list all the remote repos you have set up and their destinations.

Ramizjon commented:
@JPercivall i see this

> 
> git remote -v
> origin	https://github.com/Ramizjon/nifi.git (fetch)
> origin	https://github.com/Ramizjon/nifi.git (push)
> upstream	https://github.com/apache/nifi.git (fetch)
> upstream	https://github.com/apache/nifi.git (push)

JPercivall commented:
@Ramizjon so it looks like you have the github repo added with the name "upstream" but the current upstream repo is your "origin" repo (given that when you do "git status" it checks with origin). So try running this while on your master branch:

`git branch --set-upstream upstream`

This will set the upstream of master to point to the github repo (which you've name "upstream"). After running that, try doing a "git status"

Ramizjon commented:
@JPercivall After performing this, i'm still up-to-date, but this time with upstream.

> Your branch is up-to-date with 'upstream'

JPercivall commented:
@Ramizjon Hmmm, that is unexpected. When you do `git log` do you see "commit fa639e2596cc125af80615066f500cc908d4db6a" as the first (Matt Gilman's commit for NIFI-2542)? Sorry I thought master still had the changes you made

Ramizjon commented:
@JPercivall actually not. I see my commit first in the git log(screenshot attached)

![image](https://cloud.githubusercontent.com/assets/3027035/17626331/875b5722-60b5-11e6-9ce7-3ac42723911b.png)

JPercivall commented:
Well that is very weird. Could you do a "get fetch --all" and then "git status" and take a screenshot of the output?

Also Mac is super easy to take screen shots on, just hit `command` `shift` `4`, then you can click and drag to select what to screenshot (hitting 3 takes the entire screen). It saves it as a file to the desktop.

Ramizjon commented:
@JPercivall Yes, surely. So here is the output 
![image](https://cloud.githubusercontent.com/assets/3027035/17628584/bda927ba-60be-11e6-822f-24e80e62e752.png)

Yes, i know about that, thanks. I use Lightshot screenshot, it allows to copy to clipboard, save or instantly upload to server and retrieve a link.

JPercivall commented:
This is very odd. If you try modifying anything does the "git status" output change? I ask because, I tried searching for a possible solution and came across this SO question: http://stackoverflow.com/questions/22067873/git-nothing-to-commit-working-directory-clean

Does it sound like what is going on (up to date but have changes committed)?

Oh ok, from your pictures it looks like there are shadows so thought I'd offer the tip.

Ramizjon commented:
@JPercivall Yes, as you can see, git status output changes, depending on my modifications.

![image](https://cloud.githubusercontent.com/assets/3027035/17629234/105775ea-60c2-11e6-8cf5-158116ac551c.png)

Btw, when i was trying to squash commits i tried all approaches described here, and neither of them seemed to work..
http://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

Ah, that's probably because my terminal window is set up transparent x)

JPercivall commented:
Hmmm the only odd difference between your "git status" and mine is that mine explictly states that it is up to date with the branch of the remote.

![screen shot 2016-08-12 at 12 28 38 pm](https://cloud.githubusercontent.com/assets/11302527/17629454/6d8aa41a-6088-11e6-82cf-94356cbc1609.png)

If you do "git branch -v" do you have a branch named "upstream"?

Ah, lol that would be it.

Ramizjon commented:
@JPercivall Hm, yes i have such branch..
![image](https://cloud.githubusercontent.com/assets/3027035/17629558/a928d43e-60c3-11e6-9fe5-29a9af37b04a.png)

JPercivall commented:
Aha, let's first renaming the remote:

`git remote rename upstream github`

Then try setting it as the upstream again (added "/master" from before)

`git branch --set-upstream github/master`

Ramizjon commented:
@JPercivall doing this steps causes following error (used `--set-upstream-to` as `--set-upstream` is deprecated:

> warning: refname 'github/master' is ambiguous.
> fatal: Ambiguous object name: 'github/master'.

JPercivall commented:
Ah ok, going off of this SO question[1] try:

`git branch master --set-upstream-to github/master`

[1] http://stackoverflow.com/questions/4878249/how-do-i-change-the-remote-a-git-branch-is-tracking

Ramizjon commented:
@JPercivall still getting the same error. I've sent you an email, let me know about your decision, please

JPercivall commented:
Hey @Ramizjon, I'm actually multi-tasking on a couple of different things so can't jump on a call with you. Looking at this SO[1] try this:

`git branch master --set-upstream-to refs/remotes/github/master 

If that doesn't work I'll just work on squashing everything my self on my machine when I get a chance.

[1] http://stackoverflow.com/questions/3782893/git-fatal-ambiguous-object-name-origin-release-2-6

Ramizjon commented:
@JPercivall Yes, it finally works. I received following message (when `git status`)

> 
> On branch master
> Your branch and 'remotes/github/master' have diverged,
> and have 9 and 34 different commits each, respectively.

JPercivall commented:
@Ramizjon awesome! Now you can try resetting back 9 commits, committing just your changes, resetting to HEAD --hard. Then rebasing to github/master (may need to do the ref/ thing again).

JPercivall commented:
Also you can do "git clean -i -d" in order to clean your repo of things that aren't tracked and may have snuck in.

Ramizjon commented:
@JPercivall , what about this option? Are you not able to do that after all when doing merge?
![image](https://cloud.githubusercontent.com/assets/3027035/17667182/79040274-630c-11e6-866d-1f1c98d53b12.png)

Ramizjon commented:
@JPercivall It seems like i've succeeded to squash commits. Neither of methods didn't work, until i've tried to apply patch from PR to the necessary history point.

Ramizjon commented:
@JPercivall please let us with Ilan Rabinovitch know, if you'll be able to perform functional testing. Thanks.

JPercivall commented:
@Ramizjon I am traveling currently but yes, I will be performing basic functional testing when I am able

JPercivall commented:
I will be adding a patch on top of this to fix a couple issues with styling (see travis-ci build) and licensing but:

+1 

Visually verified the code and did a contrib check build. In a standalone instance and secure cluster I did basic functional testing in both Agent and HTTP mode (both require adding the dashboard in Datadog with the appropriate metrics). Thanks @Ramizjon for the contribution, I will merge it in.

irabinovitch commented:
Once you're ready we'd be happy to work with you all to create a tile / listing in the Datadog app, as well as a default dashboard so that users can hit the ground running with this.
